text/0000-transition-hooks-events.md

```diff
@@ -0,0 +1,122 @@
+- Animations can be handled in a complex way using `liquid-fire` but not without non-trivial performance impact for setting up new HTMLBars block contexts. The complex mechanisms rely on information gleaned from the rendering lifecycle, but a simplistic version could be implemented solely using information from the transition lifecycle if given the ability to detect modifications to the transition.
+- Knowing how and in what way to handle parallelizing data fetching requires pre-knowledge of the destination route, and updating knowledge about the destination route along the way. We can optimistically load data for the first target, and evict from the request queue if the `transition` changes.
+- Accessibility via `ember-a11y` needs knowledge of changes in transitions in order to identify when it needs to reset state to its previous container or set focus to an element responsible for loading the route.
+
+# Detailed design
+
+The `willChangeTransition` event should be fired for every new transition object created between firing `willTransition` and `didTransition`.
+
+```javascript
+const Router = Ember.Router;
+
+Router.map(function() {
+  this.route('first');
+  this.route('second');
+});
+
+Router.on('willTransition', function(transition) {
+  console.log(`transitioning to "${transition.targetName}"`);
+});
+
+Router.on('willChangeTransition', function (transition) {
+  console.log(`redirecting to "${transition.targetName}"`);
+});
+
+Router.on('didTransition', function() {
+  console.log('transition settled');
+});
+
+App.FirstRoute = Ember.Route.extend({
+  redirect() {
+    this.replaceWith('second');
+  },
+});
+
+App.SecondRoute = Ember.Route.extend();
+
+visit('/first');
+// => transitioning to "first"
+// => redirecting to "second"
+// => transition settled
+```
+
+The `willChangeTransition` hook should provide the same API surface as the `willTransition` hook, and the callback for the `willChangeTransition` event should be passed the same parameters as the callback for the `willTransition` event. This grants users wishing to subscribe to all transition events the ability to reuse one function across both hooks or both callbacks.
+
+```javascript
+function hookForAllTransitions(oldInfos, newInfos, transition) {
+  const ret = this.super(...arguments);
+
+  // do stuff for all transitions...
+
+  return ret;
+}
+
+function callbackForAllTransitions(transition) {
+  // do stuff for all transitions...
+}
+
+const Router = Ember.Router.extend({
+  willTransition: hookForAllTransitions,
+  willChangeTransition: hookForAllTransitions,
+});
+
+Router.on('willTransition', callbackForAllTransitions);
```

Comment:
Agreed. Also, `willTransition` and `didTransition` fire as actions:

http://jsbin.com/nemuzeteju/1/edit?html,js,output

It is proposed that an action be added for `willChangeTransition`? Is that possible to add, since the router is in transition (I'm not crazy familiar with the code)?